src/snowflake/snowpark/mock/_stage_registry.py

```diff
@@ -413,6 +444,73 @@ def read_file(
                 result_df = pd.concat([result_df, df], ignore_index=True)
             result_df.sf_types = result_df_sf_types
             return result_df
+        elif file_format == "json":
+            infer_schema_opt = options.get("INFER_SCHEMA", False)
+
+            result_df = TableEmulator()
+            result_df_sf_types = {}
+
+            if not infer_schema_opt:
+                # snowflake output the value as raw string
+                for i in range(len(schema)):
+                    column_name = analyzer.analyze(schema[i])
+                    column_series = ColumnEmulator(
+                        data=None, dtype=object, name=column_name
+                    )
+                    column_series.sf_type = ColumnType(
+                        schema[i].datatype, schema[i].nullable
+                    )
+                    result_df[column_name] = column_series
+                    result_df_sf_types[column_name] = column_series.sf_type
+
+                for local_file in local_files:
+                    with open(local_file) as file:
+                        content = json.load(file)
+                        df = pd.DataFrame({result_df.columns[0]: [content]})
+                        result_df = pd.concat([result_df, df], ignore_index=True)
+            else:
+                # need to infer schema
+                contents = []
+                for local_file in local_files:
+                    with open(local_file) as file:
+                        content = json.load(file)
+                        contents.append(content)
+                        # extract the schema from the content
+                        for column_name, value in content.items():
+                            target_datatype = infer_type(value)
+                            # multiple json files can be of different schema
+                            # if we find an existing schema but type is different than the inferred one
+                            # we coerce the column datatype to string
+                            if column_name in result_df_sf_types and not isinstance(
+                                target_datatype,
+                                type(result_df_sf_types[column_name].datatype),
+                            ):
+                                # snowflake coerce to string type
+                                target_datatype = StringType()
+
+                            column_series = ColumnEmulator(
+                                data=None, dtype=object, name=column_name
+                            )
+                            column_series.sf_type = ColumnType(
+                                target_datatype, nullable=True
+                            )
+                            result_df[column_name] = column_series
```

Comment:
NIT: these could be merged into one line.

Reply:
updated